### ISSUE TITLE
nvda 2021.1 beta3 fails to read messages in whatsappweb on firefox

### ISSUE THREAD

**Issue by alexandretoco** [nvaccess/nvda#12569]
### Steps to reproduce:
1. open whatsappweb in firefox.
2. Turn focus mode on
3. use tab key to find the conversations's list
4. find a conversation with arrow keys
5. press enter and tab again to find the messages's list
6. when you try to read the messages with arrow keys, it doesn't work.
7. if you press the right arrow, a context menu is opened. nvda can't also read this menu.
### Actual behavior:
nvda doesn't read messages in focus mode. it also doesn't read the messages list's context menu

### Expected behavior:
read messages and context menus correctly as it does while using schrome with whatsapp web.
#### NVDA installed/portable/running from source:
installed
#### NVDA version:
nvda 2021.1 beta3
#### Windows version:
windows 10
#### Does the issue still occur after restarting your computer?
yes
#### Have you tried any other versions of NVDA? If so, please report their behaviors.
in nvda 2020.4 it works fine
#### If add-ons are disabled, is your problem still occurring?
yes

**Comment by feerrenrut** [comment]
I can reproduce this in whatsapp. As a workaround, you can swap to browse mode to read messages.

Here is a minimal sample derived from the whatsapp code.
``` html
<button>before</button>
<div tabindex="0">
    <span>The quick brown fox.</span>
</div>
<button>after</button>
```
- Enter focus mode `NVDA+space`
- Press tab to get to the "before button"
- Press tab to read the message, "The quick brown fox" is expected.

With firefox, in NVDA 2021.1Beta3 NVDA speaks
> section

in NVDA 2020.4
> The quick brown fox

**Comment by feerrenrut** [comment]
My mistake, I hadn't changed to focus mode when testing 2020.4. With 2020.4 in focus mode using the sample above with firefox I hear "section" announced. I'm going to edit my previous message to prevent confusion.

In chrome with NVDA 2021.1Beta3 in focus mode I hear "the quick brown fox  section".

@alexandretoco Can you confirm the following two things:
- When trying to read messages in 2021.1beta3 in focus mode, you hear only "section"?
- Can you test the behavior of 2020.4 in focus mode with firefox again and confirm the behavior?

Note the behavior of Chrome is different.
While there is certainly an issue with firefox, I don't believe it is a regression / change in behavior since 2020.4. If you still think this is a regression, logs for both versions may help us get to the bottom of this.
 You can zip these and attach to this issue, or due to the sensitive nature you can email the logs to info at nvaccess.org.

**Comment by leonardder** [comment]
@jcsteh Do you have any thoughts about why this works in Chrome but not in Firefox? Seems like something at Firefox' end.

**Comment by feerrenrut** [comment]
Since this is not a regression as far as we can tell, we won't try to fix this for 2021.1

**Comment by jcsteh** [comment]
@feerrenrut, thanks for the minimal sample.

A div without an explicit role should not have a name calculated from its subtree. This is not well specified yet, but I believe the intention is to map this to role="generic", which specifies that naming is prohibited. So, this is authoring error on Whatsapp's part.

To compensate for such authoring errors, Chrome calculates the name from the subtree for divs which are focusable. We've talked about doing this in Firefox too, but this is not without its own drawbacks, particularly when you consider that scrollable regions are focusable in Firefox (to allow for keyboard scrolling), which might result in extremely long names for some divs.

So, the correct solution is for Whatsapp to fix this by giving these divs a valid ARIA role. Failing that, there's ongoing discussion about unifying behaviour across browsers (and the spec) here.

**Comment by feerrenrut** [comment]
Thanks for the explanation @jcsteh. I'm going to close this issue.

@alexandretoco Please report this to WhatsApp, you may link to this issue to help explain the cause.
